src/service/viewport/viewport-binding-ios-embed-sd.js

```diff
@@ -0,0 +1,440 @@
+
+    /** @private @const {boolean} */
+    this.useLayers_ = isExperimentOn(this.win, 'layers');
+
+    /** @private {number} */
+    this.paddingTop_ = 0;
+
+    /** @private {boolean} */
+    this.bodySyncScheduled_ = false;
+
+    // Setup UI.
+    /** @private {boolean} */
+    this.setupDone_ = false;
+    waitForBody(doc, this.setup_.bind(this));
+
+    // Set overscroll (`-webkit-overflow-scrolling: touch`) later to avoid
+    // iOS rendering bugs. See #8798 for details.
+    whenDocumentReady(doc).then(() => {
+      // There's no way to test `-webkit-overflow-scrolling` style on most of
+      // browsers. Thus we'll use the `i-amphtml-ios-overscroll` class for
+      // testing.
+      this.scroller_.classList.add('i-amphtml-ios-overscroll');
+      setImportantStyles(this.scroller_, {
+        '-webkit-overflow-scrolling': 'touch',
+      });
+    });
+
+    dev().fine(TAG_, 'initialized ios-embed-wrapper-sd viewport');
+  }
+
+  /** @override */
+  ensureReadyForElements() {
+    this.setup_();
+  }
+
+  /** @private */
+  setup_() {
+    if (this.setupDone_) {
+      return;
+    }
+    this.setupDone_ = true;
+
+    // Embedded scrolling on iOS is rather complicated. IFrames cannot be sized
+    // and be scrollable. Sizing iframe by scrolling height has a big negative
+    // that "fixed" position is essentially impossible. The only option we
+    // found is to reset scrolling on the AMP doc, which wraps the natural BODY
+    // children inside the `overflow:auto` element inside the shadow root of the
+    // body. For reference, here are related
+    // iOS issues (Chrome issues are also listed for reference):
+    // - https://code.google.com/p/chromium/issues/detail?id=2891
+    // - https://code.google.com/p/chromium/issues/detail?id=157855
+    // - https://bugs.webkit.org/show_bug.cgi?id=106133
+    // - https://bugs.webkit.org/show_bug.cgi?id=149264
+    const doc = this.win.document;
+    const body = dev().assertElement(doc.body, 'body is not available');
+    const shadowRoot = body.attachShadow({mode: 'open'});
+
+    // Main slot contained within the wrapper will absorb all undistributed
+    // children.
+    shadowRoot.appendChild(this.scroller_);
+
+    // Update body styles and monitor for further changes.
+    this.updateBodyStyles_();
+    if (this.win.MutationObserver) {
+      const bodyObserver = new this.win.MutationObserver(
+          this.updateBodyStyles_.bind(this));
+      bodyObserver.observe(body, {attributes: true});
+    }
+
+    // Make sure the scroll position is adjusted correctly.
+    this.onScrolled_();
+  }
+
+  /** @private */
+  updateBodyStyles_() {
+    const {body} = this.win.document;
+    if (!body) {
+      return;
+    }
+
+    // Many `<body>` styles are inherited to ensure that layout is preserved.
+    // The most important: `display` and related styles.
+    const inheritStyles = {};
+    this.bodySyncScheduled_ = true;
```

Comment:
There's no early return when this is already true.

Reply:
Already fixed per Will's comment above.